### PR TITLE
Added config object to engine initialization (#2057) (#5470)

### PR DESCRIPTION
**Pull request by BStringhamVRSK**
This code adds a config object, "ConfigParams," to Engine creation parameters for setting memory buffer sizes for command buffers and driver handle arena sizes.  It will use #define values (FILAMENT_PER_FRAME_COMMANDS_SIZE_IN_MB, etc.) as defaults (and as minimum acceptable values) if the user does not provide values.  It attempts to validate the values given to prevent the user from creating a unusable state.

**NOTE:** I have tried to be very careful in making changes that affect all platforms, but I am not set up to build, run, and test anything other than PC desktop at the moment.  All other platforms should build and run correctly, but they've not been directly tested.

### CODE REVIEW

**Comment by pixelflinger** [comment]
thank you for the PR, please give us a bit of time for the review

File: filament/include/filament/Engine.h [hunk 1]
```diff
@@ -54,6 +54,22 @@ class LightManager;
...
+// size of the high-level draw commands buffer (comes from the per-render pass allocator)
+#    define FILAMENT_PER_FRAME_COMMANDS_SIZE_IN_MB 2
+#endif
+
+#ifndef FILAMENT_MIN_COMMAND_BUFFERS_SIZE_IN_MB
+// size of a command-stream buffer (comes from mmap -- not the per-engine arena)
+#    define FILAMENT_MIN_COMMAND_BUFFERS_SIZE_IN_MB 1
+#endif
+
+
```
Comment (pixelflinger) [inline filament/include/filament/Engine.h:72]: I don't think we need this in the public header.

File: filament/include/filament/Engine.h [hunk 2]
```diff
@@ -151,6 +167,70 @@ class UTILS_PUBLIC Engine {
...
     using Platform = backend::Platform;
     using Backend = backend::Backend;
+    using DriverConfig = backend::Platform::DriverConfig;
+
+    /**
+    * ConfigParams are used to define the memory footprint used by the engine, such as the
+    * command buffer size. ConfigParams can be used to customize engine requirements based 
+    * on the applications needs.
+    */
+    class ConfigParams {
```
Comment (pixelflinger) [inline filament/include/filament/Engine.h:177]: Make this a simple `struct` with public fields. Also rename to `Config`. e.g.:

```
struct Config {
    uint8_t commandBufferSize = FILAMENT_COMMAND_BUFFER_SIZE_IN_MB;   //!< size of command buffer in MiB
    ...
};
```

Validate the parameters in `Engine::create`. No need to store `DriverConfig` in this structure -- just create one when needed to pass to the driver.

We basically don't want any private fields visible in public headers.

**Review by pixelflinger — CHANGES_REQUESTED** [review]
There will be more comments, but I thought I'd make those for now as it's going to be a big change already.

Thank you for the PR.

File: filament/include/filament/Engine.h (hunk 1 above)
Comment (BStringhamVRSK) [reply filament/include/filament/Engine.h:72]: Ok.  I'm using them as default values for parameters.  I will replace them with straight literals.  I'll continue to use the #define's inside the functions.

Comment (pixelflinger) [reply filament/include/filament/Engine.h:72]: Actually you're right, after you make the change where `Config` is a simple struct with default values, you'll need those in the public header -- so let's keep them here.

File: filament/backend/src/metal/MetalDriver.mm [hunk 3]
```diff
@@ -40,24 +40,30 @@
...
+Driver* MetalDriver::create(MetalPlatform* const platform, const Platform::DriverConfig& driverConfig) {
     assert_invariant(platform);
-    return new MetalDriver(platform);
+    return new MetalDriver(platform, driverConfig);
+}
+
+size_t MetalDriver::getHandleArenaSize(const Platform::DriverConfig& driverConfig) noexcept {
+    size_t configSize = driverConfig.getHandleArenaSize();
+    size_t defaultSize = FILAMENT_METAL_HANDLE_ARENA_SIZE_IN_MB * 1024U * 1024U;
+    return configSize > defaultSize ? configSize : defaultSize;
```
Comment (pixelflinger) [inline filament/backend/src/metal/MetalDriver.mm:56]: use `std::max` also document in `DriverConfig` that at least the default value will be used -- which provides a way to overrides `DriverConfig` by a build provided value.

**Review by pixelflinger — CHANGES_REQUESTED** [review]

**Comment by BStringhamVRSK** [comment]
Arghh!  I did a rebase of my changes, deleted this branch and then pushed it up again.  It closed my PR, which I did not want to happen.  Can someone reopen this PR for me?  I don't have permissions, apparently.

**Comment by BStringhamVRSK** [comment]
I pushed my branch up again, and it appears to be there.  Not sure what the problem is.

**Comment by BStringhamVRSK** [comment]
Should I create a new PR?

**Comment by romainguy** [comment]
Seems like it :/

**Comment by BStringhamVRSK** [comment]
Created new pull request: #5556

**Pull request closed without merging**